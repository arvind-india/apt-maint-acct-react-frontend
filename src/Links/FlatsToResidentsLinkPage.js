import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import {
    Input,
    Label
} from 'reactstrap'

import {
  MdHome,
  MdThumbUp,
  MdThumbDown
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { FlashMessage } from '../_components'
import { flatActions,  residentActions } from '../_actions'

let module = 'flats-residents' // module name


class FlatsToResidentsLinkPage extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      selectedOptionInLeftList: '',
      selectedOptionsInDList: [],
      selectedOptionsInAList: []
    }
    this.handleChangeInLeftList = this.handleChangeInLeftList.bind(this)
    this.handleChangeInDetachedList = this.handleChangeInDetachedList.bind(this)
    this.attachItems = this.attachItems.bind(this)
    this.handleChangeInAttachedList = this.handleChangeInAttachedList.bind(this)
    this.detachItems = this.detachItems.bind(this)
  }

  componentDidMount() {
    this.props.dispatch(flatActions.getAll())
    this.props.dispatch(residentActions.getAll())
  }

  render() {
    const { alert } = this.props
    return (
      <div>
        <h3>Flats To Residents Link</h3>
        {alert.message && <FlashMessage text={alert.message} delay={5000}/>}
        {this.show()}
      </div>
    )
  }

  show() {
    return <div className="list-to-list">
      {this.showLeftList()}
      {this.showAttachedList()}
      {this.showDetachedList()}
    </div>
  }

  showLeftList() {
    const { flats } = this.props
    return <div>
      <Label for="leftItem" className="llabel">Select a Flat</Label>
      <Input
        type="select"
        name="leftItem"
        id="leftItem"
        size="20"
        className="lselect"
        onChange={this.handleChangeInLeftList}
      >
      { flats.items && flats.items.map(each =>
        <option value={each.id} title={each.remark} key={each.id}
          >{each.block_number}-{each.flat_number}</option>)
      }
      </Input>
      <Button color="link" className="homeButton"><Link to="/home"><MdHome/> Home</Link></Button>
    </div>
  }

  handleChangeInLeftList(event) {
    const { value } = event.target
    this.setState({
      selectedOptionInLeftList: value
    })
    this.props.dispatch(flatActions.getMyResidents(value))
  }

  showAttachedList() {
    const { flatsToResidents, authzn } = this.props
    const { selectedOptionInLeftList, selectedOptionsInAList } = this.state
    return <div>
      <Label
        for="attachedItems"
        className="alabel"
      >Attached Residents</Label>
      <Input
        type="select"
        name="attachedItems"
        id="attachedItems"
        size="20"
        className="aselect"
        onChange={this.handleChangeInAttachedList}
        multiple
      >
      { selectedOptionInLeftList &&
        flatsToResidents.items &&
        flatsToResidents.items.map(each =>
        <option value={each.id} title={each.remark?each.is_a+';'+each.remark:each.is_a} key={each.id}
          >{each.first_name} {each.last_name}</option>)
      }
      </Input>
      <Button
        color="danger"
        className="dbutton"
        onClick={this.detachItems}
        disabled={selectedOptionsInAList.length === 0}
        hidden={!authzn.allowsEdit}
        title="Detach selected items"
      ><MdThumbDown/> Detach</Button>
    </div>
  }
  handleChangeInAttachedList(event) {
    const { selectedOptions } = event.target
    this.setState({
      selectedOptionsInAList: Array.from(selectedOptions)
    })
  }
  detachItems() {
    const { dispatch, flatsToResidents } = this.props
    const { selectedOptionsInAList, selectedOptionInLeftList } = this.state

    let id = selectedOptionInLeftList
    let ids_toBeRemoved = selectedOptionsInAList.map(e => e.value)
    let ids = flatsToResidents.items.map(e => e.id.toString())
    let ids_toBeRetained = ids.filter(each => !ids_toBeRemoved.includes(each))

    dispatch(flatActions.updateMyResidents(id, ids_toBeRetained))
    this.setState({ selectedOptionsInAList: [] })
  }
  showDetachedList() {
    const { flatsToResidents, residents, authzn } = this.props
    const { selectedOptionInLeftList, selectedOptionsInDList } = this.state

    let available = [];
    let grantedIDs = []
    if(flatsToResidents.items) {
      grantedIDs = flatsToResidents.items.map(each => each.id)
    }
    if(residents.items && selectedOptionInLeftList) {
      available = grantedIDs.length ?
        residents.items.filter(each => !grantedIDs.includes(each.id)) :
        residents.items
    }
    return <div>
      <Label
        for="detachedItems"
        className="dlabel"
      >Available Residents</Label>
      <Input
        type="select"
        name="detachedItems"
        id="detachedItems"
        size="20"
        className="dselect"
        onChange={this.handleChangeInDetachedList}
        multiple
      >
      {
        available.map(each =>
        <option value={each.id} title={each.remark?each.is_a+';'+each.remark:each.is_a} key={each.id}
          >{each.first_name} {each.last_name}</option>)
      }
      </Input>
      <Button
        color="success"
        className="abutton"
        onClick={this.attachItems}
        disabled={selectedOptionsInDList.length === 0}
        hidden={!authzn.allowsEdit}
        title="Attach selected items"
      ><MdThumbUp/> Attach</Button>
    </div>
  }

  handleChangeInDetachedList(event){
    const { selectedOptions } = event.target
    this.setState({
      selectedOptionsInDList: Array.from(selectedOptions)
    })
  }
  attachItems() {
    const { dispatch, flatsToResidents } = this.props
    const { selectedOptionsInDList, selectedOptionInLeftList } = this.state

    let id = selectedOptionInLeftList
    let ids = flatsToResidents.items.map(e => e.id)
    let ids_toBeAdded = selectedOptionsInDList.map(e => e.value)
    let new_ids = ids.concat(ids_toBeAdded)

    dispatch(flatActions.updateMyResidents(id, new_ids))
    this.setState({ selectedOptionsInDList: [] })
  }
}

function mapStateToProps(state) {
  const { alert, flats, flatsToResidents, residents, authorizations } = state
//  const { user } = authentication
  const authzn = authorizations[module]
  return {
    alert,
//    user,
    flats,
    flatsToResidents,
    residents,
    authzn
  }
}

const connectedLinkPage = connect(mapStateToProps)(FlatsToResidentsLinkPage)
export { connectedLinkPage as FlatsToResidentsLinkPage }
