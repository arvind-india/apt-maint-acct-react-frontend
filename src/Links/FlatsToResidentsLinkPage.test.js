// FlatsToResidentsLinkPage.test.js0]

import React from 'react'
import { shallow } from 'enzyme'

import { FlatsToResidentsLink } from './FlatsToResidentsLinkPage'

const mockFlats = [
  { id: 1, block_number: '0', flat_number: 'G1', remarks: 'test on G1' },
  { id: 2, block_number: '0', flat_number: 'G2', remarks: 'test on G2' }
]

const mockResidents = [
  { id: 1, first_name: 'user1', last_name: 'test', is_a: 'owner', remark: 'test user1' },
  { id: 2, first_name: 'user2', last_name: 'test', is_a: 'owner', remark: 'test user2' }
]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    flats: { items: mockFlats },
    flatsToResidents: { items: [] },
    residents: { items: mockResidents },
    getAllFlats: jest.fn(),
    getAllResidents: jest.fn(),
    getMyResidents: jest.fn(),
    updateMyResidents: jest.fn()
  },
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    getAllFlats: jest.fn(),
    getAllResidents: jest.fn(),
    getMyResidents: jest.fn(),
    updateMyResidents: jest.fn()
  }
]

const component = shallow(<FlatsToResidentsLink {...mockProps[0]} />)

describe('FlatsToResidentsLinkPage params test', () => {
  it('should render FlatsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionInLeftList')).toBeDefined()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionsInDList')).toBeDefined()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionsInAList')).toBeDefined()
  })
})

describe('FlatsToResidentsLinkPage test on selections, attach, and detach', () => {

  component.find('#leftItem').simulate('change', {target: {name:'id', value: mockFlats[0].id}})

  it('should check options in Left List', () => {
    expect(component.state('selectedOptionInLeftList')).toEqual(mockFlats[0].id)
  })

  component
    .find('#detachedItems')
    .simulate('change', {
                          target: {
                            name: 'id',
                            selectedOptions: [mockResidents[0].id, mockResidents[1].id]
                          }
                        }
    )
  it('should check options in detached list', () => {
    expect(component.state('selectedOptionsInDList'))
      .toEqual([mockResidents[0].id, mockResidents[1].id])
  })
})
