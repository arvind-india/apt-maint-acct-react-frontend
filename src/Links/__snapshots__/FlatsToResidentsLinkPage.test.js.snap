// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatsToResidentsLinkPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Flats To Residents Link
    </h3>
    <FlashMessage
        delay={5000}
        text="testingAlert"
    />
    <div
        className="list-to-list"
    >
        <Label
            className="llabel"
            for="leftItem"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Select a Flat
        </Label>
        <Input
            className="lselect"
            id="leftItem"
            name="leftItem"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            <option
                title={undefined}
                value={1}
            >
                0
                -
                G1
            </option>
            <option
                title={undefined}
                value={2}
            >
                0
                -
                G2
            </option>
        </Input>
        <div
            class="homeButton"
        >
            <Button
                color="link"
                tag="button"
            >
                <Link
                    replace={false}
                    to="/home"
                >
                    <MdHome />
                     Home
                </Link>
            </Button>
        </div>
        <Label
            className="alabel"
            for="attachedItems"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Attached Residents
        </Label>
        <Input
            className="aselect"
            id="attachedItems"
            multiple={true}
            name="attachedItems"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            
        </Input>
        <Button
            className="dbutton"
            color="danger"
            disabled={true}
            hidden={false}
            onClick={[Function]}
            tag="button"
            title="Detach selected items"
        >
            <MdThumbDown />
             Detach
        </Button>
        <Label
            className="dlabel"
            for="detachedItems"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Available Residents
        </Label>
        <Input
            className="dselect"
            id="detachedItems"
            multiple={true}
            name="detachedItems"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            <option
                title="owner;test user1"
                value={1}
            >
                user1
                 
                test
            </option>
            <option
                title="owner;test user2"
                value={2}
            >
                user2
                 
                test
            </option>
        </Input>
        <Button
            className="abutton"
            color="success"
            disabled={true}
            hidden={false}
            onClick={[Function]}
            tag="button"
            title="Attach selected items"
        >
            <MdThumbUp />
             Attach
        </Button>
    </div>
</div>,
]
`;
